cleanrl/cql_continuous/sac_cql.py

```diff
@@ -0,0 +1,396 @@
+        values_optimizer.step()
+
+        ## Policy network update
+        if global_step % args.policy_frequency == 0: # TD 3 Delayed update support
+            for _ in range(args.policy_frequency): # compensate for the delay by doing 'actor_update_interval' instead of 1
+                pi, log_pi, _ = pg.get_action(s_obs)
+                qf1_pi = qf1.forward(s_obs, pi)
+                qf2_pi = qf2.forward(s_obs, pi)
+                min_qf_pi = torch.min(qf1_pi, qf2_pi).view(-1)
+                policy_loss = ((alpha * log_pi) - min_qf_pi).mean()
+
+                policy_optimizer.zero_grad()
+                policy_loss.backward()
+                policy_optimizer.step()
+            
+            if args.autotune:
+                with torch.no_grad():
+                    _, log_pi, _ = pg.get_action(s_obs)
+                alpha_loss = ( -log_alpha * (log_pi + target_entropy)).mean()
+
+                a_optimizer.zero_grad()
+                alpha_loss.backward()
+                a_optimizer.step()
+                alpha = log_alpha.exp().item()
+            
+        # update the target network
+        if global_step % args.target_network_frequency == 0:
+            for param, target_param in zip(qf1.parameters(), qf1_target.parameters()):
+                target_param.data.copy_(args.tau * param.data + (1 - args.tau) * target_param.data)
+            for param, target_param in zip(qf2.parameters(), qf2_target.parameters()):
+                target_param.data.copy_(args.tau * param.data + (1 - args.tau) * target_param.data)
+        
+        # Stop iteration over the dataset
+        if k >= args.epoch_length-1:
+            break
+    
+    eval_stats = test_agent(env, pg)
+    print("[E%4d|I%8d] -- PLoss: %.3f -- QLoss: %.3f -- CQLQloss: %.3f --  Train Mean Ret: %.3f" %
+        (epoch, global_step, policy_loss.item(), qf_loss.item(), cql_qf_loss.item(),
+        eval_stats["test_mean_return"]))
+    if args.with_lagrange:
+        print("\t APrimeLoss: %.3f" % alpha_prime_loss.item())
+
+    writer.add_scalar("losses/soft_q_value_1_loss", qf1_loss.item(), global_step)
+    writer.add_scalar("losses/soft_q_value_2_loss", qf2_loss.item(), global_step)
+    writer.add_scalar("losses/qf_loss", qf_loss.item(), global_step)
+    writer.add_scalar("losses/cql_qf_loss", cql_qf_loss.item(), global_step)
+    writer.add_scalar("losses/policy_loss", policy_loss.item(), global_step)
+    writer.add_scalar("losses/alpha", alpha, global_step)
+    writer.add_scalar("charts/episode_reward", eval_stats["test_mean_return"], global_step)
+    writer.add_scalar("global_step", global_step, global_step)
+    if args.autotune:
+        writer.add_scalar("losses/alpha_loss", alpha_loss.item(), global_step)
+    if args.with_lagrange:
+        writer.add_scalar("losses/alpha_prime_loss", alpha_prime_loss.item(), global_step)
+        writer.add_scalar("losses/alpha_prime", log_alpha_prime.exp().item(), global_step)
+
+writer.close()
+env.close()
```

Comment:
please add a newline in the end

Reply:
Done. Quite curious about the reason for the last line though ?